### PR TITLE
Add PSCI driver

### PR DESCRIPTION
**Pull request by carlocaione**
A bit of rationale behind this PR.

@MrVan submitted in #30676 the PR to enable SMP on AArch64. As I wrote in a review to that PR I do not particularly like the approach of using arch-specific code to interface with the PSCI firmware.

PSCI interface needs a proper driver that uses the DTS to detect the correct parameters, especially the conduit (and currently only that). This is particularly important because the choice of having SMC or HVC as conduit really depends on whether we are running an hypervisor or not in EL2, so this must be something configurable at DTS (board) level.

So here it is a proper PSCI driver with support  for PSCI 0.2 and PSCI 1.0, enabled using the DT and also a tiny test to make sure that everything is working correctly.

### CODE REVIEW

File: arch/arm/core/aarch64/smccc-call.S [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2019 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <arch/cpu.h>
+
```
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/smccc-call.S:10]: Just a note here would help someone unfamiliar to the fact that it implements a spec. 
The SMCCC  excerpt to show what parameter means what would also help :) .  Being greedy here.

**Review by sandeepbrcm — APPROVED** [review]

File: arch/arm/core/aarch64/smccc-call.S (hunk 1 above)
Comment (carlocaione) [reply arch/arm/core/aarch64/smccc-call.S:10]: you are totally right in being greedy here ;) I'll add some comments.

Comment (carlocaione) [reply arch/arm/core/aarch64/smccc-call.S:17]: @sandeepbrcm I added some comments and a link to the spec. Let me know if you want me to add anything else.

**Review by MrVan — APPROVED** [review]

**Comment by MrVan** [comment]
@carlocaione Thanks for adding psci support, your code looks good to me. Approved.

**Comment by carlocaione** [comment]
Can we merge this?

**Review by MaureenHelm — APPROVED** [review]

File: tests/arch/arm/arm_psci/src/main.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	ver = psci_get_version(psci);
+	zassert_false((PSCI_VERSION_MAJOR(ver) == 0 &&
+		       PSCI_VERSION_MINOR(ver) < 2),
+		       "Wrong PSCI firware version");
+}
+
+void test_main(void)
+{
+	ztest_test_suite(psci_func,
+		ztest_unit_test(test_psci_func));
```
Comment (andrewboie) [inline tests/arch/arm/arm_psci/src/main.c:29]: suggest running this in userspace with ztest_user_unit_test(), granting permission in test_main beforehand

**Review by tyrantlizard77 — COMMENTED** [review]
Overall looks great but it would be better to have code coverage of the syscalls by running the test function in user mode, and have the test exercise all of the APIs.

**Comment by carlocaione** [comment]
> Overall looks great but it would be better to have code coverage of the syscalls by running the test function in user mode, and have the test exercise all of the APIs.

@andrewboie I addressed your points. The only API I couldn't test was `psci_cpu_off` (because that actually shuts the CPU off).

File: tests/arch/arm/arm_psci/src/main.c (hunk 2 above)
Comment (carlocaione) [reply tests/arch/arm/arm_psci/src/main.c]: Done in the new revision

**Pull request merged**